### ISSUE TITLE
TimeTable

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#472]
**Reported by petar on 21 Dec 2010 11:48 UTC**
The current implementation of TimeTable has both this declaration

```
 SIunits.Time nextEvent(start=0, fixed=true) "Next event instant";
```
and this when clause:

```
algorithm
when {time >= pre(nextEvent),initial()} then
      (a,b,nextEvent,last):=getInterpolationCoefficients(table, offset, startTime, time, last, 100*Modelica.Constants.eps);
  end when;
```

Since the when clause has an initial() present, it means that the algorithm is "active" during initialization, thus the initialization is overdetermined. Either, initial() should be removed or fixed=true.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/472

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 3 Jun 2012 11:26 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 19 Sep 2012 11:19 UTC**
3.8.3 refers to a discrete-time variable as "Integer, Boolean, String variables and enumeration variables, as well as Real variables assigned in when-clauses"

As pointed out in #824, 3.6 states "For all discrete variables vd, the equation pre(vd) = startExpression is added to the initialization
equations, if start = startExpression and fixed = true"

For the purposes of initialization is nextEvent a discrete variable due to it's assignment in a when-clause, and therefore the "start=0, fixed=true" modification actually denotes "pre(nextEvent) = 0" (and hence not overdetermined)? It is certainly discrete in the simulation but it is not declared as discrete.

The definition of the contents of "when initial()..." being "active" at initialization isn't entirely clear, the obvious assumption is that it means the contents are equivalent to an initial equation, however what is the result when it is in an algorithm clause? The following example (Modelica.Electrical.Digital.Sources.Step) has a "when initial()..." and an algorithm to do the same thing (and in the case of stepTime==0.0 actually gives a conflicting result):

```
  block Step "Digital Step Source"
      import D = Modelica.Electrical.Digital;
      import L = Modelica.Electrical.Digital.Interfaces.Logic;
      parameter D.Interfaces.Logic before(start=L.'0') "Logic value before step";
      parameter D.Interfaces.Logic after(start=L.'1') "Logic value after step";
      parameter Real stepTime(start=1) "step time";
      D.Interfaces.DigitalOutput y annotation (Placement(transformation(extent={{90,-10},{110,10}}, rotation=0)));
        algorithm
          when initial() then
            y := before;
          end when;
          if time >= stepTime then
            y := after;
          else
            y := before;
          end if;
  end Step;
```

**Comment by modelica-trac-importer** [comment]
**Changelog removed by pharman on 19 Sep 2012 11:19 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 24 Sep 2012 13:15 UTC**
Fixed in 0519425bda7f3a70d81cb2b31edecf3566a4481d, by explicitly declaring that nextEvent in TimeTable is discrete and therefore unambiguous as to how it is initialized, and removing "when initial() ..." clause from Step.
